### DISCUSSION TITLE
ContentReceptor? PasteTarget? Placeholder control for receiving paste/drag-n-drop text/image content

### DISCUSSION THREAD

**Discussion by michael-hawker** [CommunityToolkit/Labs-Windows#499] [Ideas]
# Problem Statement

A relatively common task (albeit difficult to implement properly across UWP and WinUI 3 apps) is allowing users to paste content from their clipboard, especially an image, into an application. Or drag-n-drop a file into the app as well (or from another location in their app).

## Idea

![image](https://github.com/CommunityToolkit/Labs-Windows/assets/24302614/61e11ce6-8b4b-41d0-bcb0-45304ed94ca3)

By having an easy drop-in-place control that can represent a target area for a user to drag a file to or paste an image into, it can enable a wide-variety of scenarios and input into an application, while abstracting away the nuances of how to get various data from  different places.

The control should provide a nice default style and look-and-feel, but also be flexible to encompass being used as a wrapper around other content and directing content to other locations (possibly via events that the developer uses?)

## Open Questions/Thoughts

- Is there more beyond text, images, files?
- Bools for allowing certain operations like `AllowDrop` and `AllowPaste`?
- Some sort of enum or boolean properties for `AllowImage`, `AllowText`, etc... Maybe better to have as an enum for allowed content types? Maybe they're flags so it can be `AllowableContent="Image,Text"`?
- Have events like `ReceivedImage`, `ReceivedText`, `ReceivedFile` which just provides nice callbacks when content is pasted or dragged (i.e. 'received') by the control and provides the content directly as a string, `BitmapImage`, and `StorageFile`?
- Should events be intercept-able/handle-able or is allow filters + post-events enough?
- How much text do we display, do we truncate? Is there an option to show as a 'file icon' (do we have different images there or is that to the developer...)?
- Option/Add-on to have button and open file picker to select a file as well, trifecta paste/drag-n-drop/filepicker... 🙂

In general, I like the idea of this control and see a lot of potential, but think there's a fine-line in scoping we need to hit in aggregating these types of operations and simplifying in/out-of-app communication and preview/display of the result, etc...

**Comment by niels9001** [comment]
Simple Icon File Maker has a UI that could benefit from this idea!

![apps.8080.14377116441060232.f6e53940-1b75-4b00-939e-ddf86bc5d5e9.1c713fee-3213-4669-9357-6a39af36271d.jpg](https://github.com/CommunityToolkit/Labs-Windows/assets/9866362/bfa7f3f2-a837-463d-aa08-d3cb4ad63d93)

**Reply by michael-hawker** [reply]
Thanks for the example Niels, yeah, this kind of placeholder UI is in many apps. Forgot to call out the button to select a file/image as well, that would be another add-on/option to the component.

**Comment by michael-hawker** [comment]
Link to some API docs: https://learn.microsoft.com/en-us/windows/uwp/app-to-app/

![image](https://github.com/CommunityToolkit/Labs-Windows/assets/24302614/2432486a-a2db-4119-8b36-37d36a429bb1)

Wonder if the share out/in contracts could be useful to integrate somewhere here as well? @niels9001?
